cmd/nirvana/project/init.go

```diff
@@ -0,0 +1,782 @@
+{{- if .Boilerplate -}}
+{{ .GoBoilerplate }}
+{{- end }}
+
+package descriptors
+
+import (
+	"{{ .ProjectPackage }}/pkg/api/v1/middlewares"
+
+	def "github.com/caicloud/nirvana/definition"
+)
+
+// descriptors describe APIs of current version.
+var descriptors []def.Descriptor
+
+// register registers descriptors.
+func register(ds ...def.Descriptor) {
+	descriptors = append(descriptors, ds...)
+}
+
+// Descriptor returns a combined descriptor for current version.
+func Descriptor() def.Descriptor {
+	return def.Descriptor{
+		Description: "v1 APIs",
+		Path:        "/v1",
+		Middlewares: middlewares.Middlewares(),
+		Children:    descriptors,
+	}
+}
+`
+}
+
+func (o *initOptions) templateFilters() string {
+	return `
+{{- if .Boilerplate -}}
+{{ .GoBoilerplate }}
+{{- end }}
+
+package filters
+
+import "github.com/caicloud/nirvana/service"
+
+// Filters returns a list of filters.
+func Filters() []service.Filter {
+	return []service.Filter{
+		service.RedirectTrailingSlash(),
+		service.FillLeadingSlash(),
+		service.ParseRequestForm(),
+	}
+}
+`
+}
+
+func (o *initOptions) templateMiddlewares() string {
+	return `
+{{- if .Boilerplate -}}
+{{ .GoBoilerplate }}
+{{- end }}
+
+package middlewares
+
+import def "github.com/caicloud/nirvana/definition"
+
+// Middlewares returns a list of middlewares.
+func Middlewares() []def.Middleware {
+	return []def.Middleware{}
+}
+`
+}
+
+func (o *initOptions) templateModifiers() string {
+	return `
+{{- if .Boilerplate -}}
+{{ .GoBoilerplate }}
+{{- end }}
+
+// +nirvana:api=modifiers:"Modifiers"
+
+package modifiers
+
+import "github.com/caicloud/nirvana/service"
+
+// Modifiers returns a list of modifiers.
+func Modifiers() []service.DefinitionModifier {
+	return []service.DefinitionModifier{
+		service.FirstContextParameter(),
+		service.ConsumeAllIfConsumesIsEmpty(),
+		service.ProduceAllIfProducesIsEmpty(),
+		service.ConsumeNoneForHTTPGet(),
+		service.ConsumeNoneForHTTPDelete(),
+		service.ProduceNoneForHTTPDelete(),
+	}
+}
+`
+}
+
+func (o *initOptions) templateAPI() string {
+	return `
+{{- if .Boilerplate -}}
+{{ .GoBoilerplate }}
+{{- end }}
+
+// +nirvana:api=descriptors:"Descriptor"
+
+package api
+
+import (
+	"{{ .ProjectPackage }}/pkg/api/middlewares"
+	v1 "{{ .ProjectPackage }}/pkg/api/v1/descriptors"
+
+	def "github.com/caicloud/nirvana/definition"
+)
+
+// Descriptor returns a combined descriptor for APIs of all versions.
+func Descriptor() def.Descriptor {
+	return def.Descriptor{
+		Description: "APIs",
+		Path:        "/api",
+		Middlewares: middlewares.Middlewares(),
+		Consumes:    []string{def.MIMEJSON},
+		Produces:    []string{def.MIMEJSON},
+		Children: []def.Descriptor{
+			v1.Descriptor(),
+		},
+	}
+}
+`
+}
+
+func (o *initOptions) templateMessageAPI() string {
+	return `
+{{- if .Boilerplate -}}
+{{ .GoBoilerplate }}
+{{- end }}
+
+package descriptors
+
+import (
+	"{{ .ProjectPackage }}/pkg/message"
+
+	def "github.com/caicloud/nirvana/definition"
+)
+
+func init() {
```

Comment:
I'm a little concerned about using init() to register handlers, is there a better alternative?

Reply:
:point_up_2: what about this?